### DISCUSSION TITLE
Model accessor as column field (V2)

### DISCUSSION THREAD

**Discussion by cytech** [rappasoft/laravel-livewire-tables#831] [Q&A]
In V1 , I could get the model accessor as column data.
In V2  it throws sql errors that the column does not exist.
Am I missing something or is this no longer possible?

where 'status_text' is an accessor in the model:
```
Column::make(__('bt.status'), 'status_text')
                    ->format(function ($value, $row, Column $column) use ($statuses) {
                        $ret = '<span class="badge badge-' . strtolower($value) . '">' . $statuses[$value] . '</span>';
                        if ($row->viewed)
                            $ret .= '<span class="badge bg-success">' . trans('bt.viewed') . '</span>';
                        else
                            $ret .= '<span class="badge bg-secondary">' . trans('bt.not_viewed') . '</span>';
                        return $ret;
                    })
                    ->html(),
```

model accessor:

```
public function getStatusTextAttribute()
{
$statuses = QuoteStatuses::statuses();

        return $statuses[$this->attributes['quote_status_id']];
    }
```

and it exists in $appends:
```
protected $appends = ['formatted_quote_date', 'formatted_expires_at','status_text', 'formatted_summary'];
```

**Comment by elsoftpy** [comment]
Perhaps this helps:

In the model 
```
public function isActive(): Attribute
{
     return Attribute::make(
         get: fn($value) => $this->active ? 'Yes' : 'No',
     );
}
```
in the tableComponent
```
public function columns(): array
{
     return[
          Column::make('Active', 'active')
          ->format(
             fn($value, $row, Column $column) => $row->isActive
          ), 
     ];
}
```

Taken from the [docs](https://rappasoft.com/docs/laravel-livewire-tables/v2/columns/available-methods#content-modifying-the-column-value)

**Reply by cytech** [reply]
yes that will work (kind of what i ended up doing).
it just means duplicating (or changing) a lot of logic in the existing code as multiple accessors are accessed everywhere...
thanx

**Reply by nasrulhazim** [reply]
> ->format(
>              fn($value, $row, Column $column) => $row->isActive
>           ),

On the side note, if your accessor is doing something with relationships, then you need to include the `id`. 

**This only applicable if you are using different primary key, but your relationship still rely on standard Laravel convention.**

I have the accessor:

```php
    public function getSignInCountAttribute()
    {
        return $this->authentications()->count();
    }
```

And to load it properly, I have to do as following, else I always get zero.

```php
    public function configure(): void
    {
        $deparment = Department::with('organization')->whereUuid($this->uuid)->firstOrFail();

        $this->setPrimaryKey('uuid')
            ->setAdditionalSelects(['users.id'])
            ->setConfigurableAreas([
                'before-toolbar' => ['employees.partials.create-modal', [
                    'department' => $deparment,
                ]],
            ]);
    }
```

Am not sure if this approach is appropriate. My concern is not reveal the Primary Key / Foreign Key at all. I always use UUID as my external reference. PK, FK always use at the database level.

**Comment by imabug** [comment]
I ended up using `label()` and re-querying my model to get the model attribute I wanted

`Column::make('Foo')->label(fn ($row) => Model::find($row->id)->attribute)`

Everything else I tried just gave me SQL errors about non-existent columns or nothing at all.

**Comment by nathan-io** [comment]
You shouldn't have to re-query the model.

The key is to ensure that any columns relied upon by your accessor are either defined as a column on the table, or manually included in the query via `setAdditionalSelects()`.

What worked for me:

In `configure()`

```
        $this->setAdditionalSelects([
            'items.id as id',
            'items.fine_weight as fine_weight',
            'items.weighing_unit_type as weighing_unit_type',
        ]);
```

In `columns()`:

```
            Column::make('Fine Content')
                ->label(
                    fn($row, Column $column) => $row->fine_weight_formatted
                ),
```

**Comment by lrljoe** [comment]
Closing due to age.  Appreciate you responding to discussions, but this one was a year old!

**Comment by nathan-io** [comment]
I spent quite a bit of time on this issue. `$row` was always `null` because we didn't have a column for `id`. 

It may be a good idea to put a warning in the docs that you should use `setAdditionalSelects()` to get the `id` if you don't have a column for it, else things may not work as expected. 

Apologies if they already have something about this, I don't remember seeing it.

**Reply by lrljoe** [reply]
I do remember writing about this, but will double check that it hasn't been overwritten in a subsequent update.

It is definitely in the docs:

![Screenshot_20230816-161157.png](https://github.com/rappasoft/laravel-livewire-tables/assets/104938042/4800d562-d9db-421a-ba13-e633263b098e)
